src/a.tex

```diff
@@ -157,29 +159,49 @@ \section{Load-Reserved/Store-Conditional Instructions}
 Emulating misaligned LR/SC sequences is impractical in most systems.
 \end{commentary}
 
-An implementation can register an arbitrarily large reservation set on each
-LR, provided the reservation set includes all bytes of the addressed data word
-or doubleword.
-An SC can only pair with the most recent LR in program order.
-An SC may succeed if no store from another hart, nor a write from some other
-device, to the reservation set can be observed to have occurred between the LR
-and the SC, and if there is no other SC between the LR and itself in program
-order.
-Note this LR might have had a different address argument and data size, but
-reserved the SC's address as part of the reservation set.
-Following this model, in systems with memory translation, an SC is allowed to
-succeed if the earlier LR reserved the same location using an alias with
-a different virtual address, but is also allowed to fail if the virtual
-address is different.
-The SC must fail if the address is not within the reservation set of the most
-recent LR in program order.
-The SC must fail if a store from another hart, or a write from some other
-device, to the reservation set can be observed to occur between the LR and the
-SC.
-An SC must fail if there is another SC (to any address) between the LR and the
-SC in program order.
-The precise statement of the atomicity requirements for successful LR/SC
-sequences is defined by the Atomicity Axiom in Section~\ref{sec:rvwmo}.
+The reservation set accessed by the LR and SC instructions must be {\em
+reservable}, i.e., the addresses in the reservation set must support these
+operations; otherwise, executing an LR or SC instruction generates an access
+exception.
+
+An LR registers an arbitrarily large reservation set, and the reservation set
+must include all bytes of the addressed data word or doubleword.  An SC
+succeeds if all of the following conditions are true when the SC is executed:
+\vspace{-0.2in}
+\begin{itemize}
+\parskip 0pt
+\itemsep 1pt
+\item The reservation is valid.
+\item The reservation set includes all of the bytes of the word or doubleword
+addressed by the SC.
+\end{itemize}
+If any of the above conditions are false when the SC is executed, the SC fails.
+
+The following events are guaranteed to invalidate the reservation on hart
+{\em H}:
+\vspace{-0.2in}
+\begin{itemize}
+\parskip 0pt
+\itemsep 1pt
+\item {\em H} executes a successful SC.
+\item {\em H} executes an SC that addresses bytes that are not included in the
+reservation set.
```

Comment:
The reservation should be invalidated if any SC instruction is executed, and it's not clear to me these two bullet points are collectively exhaustive of SC instructions, e.g. because SCs can fail for no particular reason from time to time.  I would merge them into a single item: "{\em H} executes an SC instruction (irrespective of the success of the SC instruction)."

Reply:
We were trying to list the things that were guaranteed to invalidate the reservation across all implementations of the architecture. I intentionally omitted a failed SC because the reservation would already be invalid in that case.

Reply:
The two bullets look correct.  The single bullet is simpler and seems clearer -- it's okay to invalidate a reservation that's already invalid.  The distinction of the 2 bullets is perhaps most relevant in showing that a constrained LR/SC sequence never itself invalidates the reservation except by succeeding -- material for non-normative comments?

@aswaterman, as defined so far, SC cannot fail for no particular reason from time to time, unless saying that the reservation can be cleared for no particular reason from time to time (which would be another bullet in this list).

Reply:
OK. (in my best NPR car talk voice) though Andrew cringes every time he hears me say it, I agree with him on this one. 

SC should clear a reservation no matter what (including it already being cleared). 

First off, it's legal to fail an SC even if the address and size match and the reservation is still standing if the sequence is non-conforming ("LR/SC sequences that do not lie withing constrained LR/SC loops are unconstrained. Unconstrained LR/SC sequences might succeed on some attempts on some implementations, but might never succeed on other implementations"). Hence you cannot assume the reservation is absent if the SC fails.

Second, even if it were the case that you know the reservation was gone on a failing SC, I would still argue that it's better to be explicit that the reservation is always gone after an SC. Don't require the reader to make the leap of though that ... oh, a failing SC, so the reservation must be down, so that's why were skpping it here in the architecture text. It more robust and less prone to mental failure to be explicit.

Finally, it does no harm to reset an already reset reservation.

I would stick with SC always kill the reservation (even dead ones).

@marceg, I think SC can fail for no particular reason... at least in non-normative text (that if not supported by normative text should be made normative text). To wit, the non-normative text says "These definitions admit the possibility that SC instructions may spuriously fail for for [someone should remove the second for :-)] implementation reasons, provided progress is eventually made."

This was a change I asked for specifically to ensure that an implementation that had a spurious rarely occurring glitch or simply a rarely occurring case where they had to fail an SC that would otherwise pass, they could without being held up as "broken". You don't want to be absolute here on the must pass when spurious fails that don't impede progress do no harm and really don't have  to be fixed.

Reply:
OK. My meaning was to specify what events cause the reservation to transition from valid to invalid, but I think what folks are asking for is more along the lines of "after these events, the reservation is guaranteed to be invalid."

So here's my proposal for the bullets and surrounding text:

```
The reservation is guaranteed to be invalid after the following events occur:
- H executes an SC instruction
- A memory write (e.g., an unconditional store, an AMO, a successful SC, or a device write) performed by another hart or device to a byte included in the reservation set is observed between the LR and the SC executed by H

The reservation may be invalid after implementation-defined events occur, provided the invalidation does not impede the eventual success guarantee in Section 8.3. In addition, architectural specifications that extend the base specification, e.g. extensions, privileged architecture, and platform specifications, may identify additional events that result in an invalid reservation; however, the specification of these events may change the conditions under which the guarantee holds.
```

Or something like that...